### PR TITLE
AA-606 Updated savings planner formula

### PR DESCRIPTION
**Pull request by brumik**
+ added basic eslint rules for typescript as they were missing.
I accidentally pushed to the same commit, if it is a problem I can split it up.

### CODE REVIEW

File: src/Components/ModalContents.js [hunk 1]
```diff
@@ -247,7 +247,7 @@ const ModalContents = ({ selectedId, isOpen, handleModal, qp, jobType }) => {
   return (
     <Modal
       aria-label="modal"
-      width={'80%'}
+      variant={ModalVariant.medium}
```
Comment (brumik) [inline src/Components/ModalContents.js:250]: Extra change to stay consistent with the modal width.

File: src/Containers/SavingsPlanner/Details/StatisticsTab/index.tsx [hunk 2]
```diff
@@ -251,59 +259,81 @@ const StatisticsTab: FunctionComponent<Props> = ({ tabsArray, data }) => {
...
-        </GridItem>
+        <GridItem span={9}>{renderLeft()}</GridItem>
+        <GridItem span={3}>{renderRight()}</GridItem>
       </Grid>
     </Card>
   );
 };
 
+StatisticsTab.propTypes = {
+  /* eslint-disable-next-line */
```
Comment (brumik) [inline src/Containers/SavingsPlanner/Details/StatisticsTab/index.tsx:327]: The validation for the data was crazy, and I gave up halfway, so I disabled the check (without ts-ignore it does not compile)

File: package.json [hunk 3]
```diff
@@ -59,6 +59,7 @@
     "@testing-library/react-hooks": "^5.1.2",
     "@testing-library/user-event": "^13.1.8",
     "@types/styled-components": "^5.1.9",
+    "@typescript-eslint/eslint-plugin": "^4.28.2",
```
Comment (jlmitch5) [inline package.json:62]: i think it's easier if stuff like this gets pulled into a separate pr, but it's fine, and good that we are doing it, nice stuff.

File: src/Api/api.ts [hunk 4]
```diff
@@ -45,27 +51,57 @@ export const getFeatures = async () => {
...
+): Promise<ApiJson> => get(planOptionsEndpoint, params);
 
-export const readClusters = () => get(clustersEndpoint);
-export const readClustersOptions = ({ params }: ReadParams) => post(clustersOptionsEndpoint, params);
+export const readClusters = (): Promise<ApiJson> => get(clustersEndpoint);
+export const readClustersOptions = ({ params }: ReadParams): Promise<ApiJson> =>
+  post(clustersOptionsEndpoint, params);
 
-export const readNotifications = ({ params }: ReadParams) => get(notificationsEndpoint, params);
+export const readNotifications = ({ params }: ReadParams): Promise<ApiJson> =>
```
Comment (jlmitch5) [inline src/Api/api.ts:106]: I believe this file has just gotten a type script make over and there are no functionality changes.

File: src/Api/methods.ts [hunk 5]
```diff
@@ -65,14 +83,18 @@ export const postWithPagination = (endpoint: string, params: ParamsWithPaginatio
...
     method: 'DELETE',
   }).then(handleResponse);
 };
 
-export const updateById = (endpoint: string, id: string, params: Params = {}) => {
+export const updateById = (
+  endpoint: string,
+  id: string,
+  params: Params = {}
+): Promise<ApiJson> => {
```
Comment (jlmitch5) [inline src/Api/methods.ts:97]: Same with this one.

File: src/Api/types.ts [hunk 6]
```diff
@@ -1,13 +1,14 @@
...
+  sort_by?: string;
+  [x: string]: string | number | undefined;
+}
 
 export type ReadParams = { params: Params };
 export type ReadParamsWithPagination = { params: ParamsWithPagination };
 export type DeleteParams = { id: string };
-export type UpdateParams = { id: string, params: Params };
\ No newline at end of file
+export type UpdateParams = { id: string; params: Params };
```
Comment (jlmitch5) [inline src/Api/types.ts:14]: Again, all this api refactoring is great stuff, but it does overload the goal of this pr (which is adding the new savings planner formula modal).

File: src/Components/ModalContents.js (hunk 1 above)
Comment (jlmitch5) [reply src/Components/ModalContents.js:250]: Nice!

File: src/Components/Pagination/Pagination.tsx [hunk 7]
```diff
@@ -34,10 +39,10 @@ const Pagination: FunctionComponent<Props> = ({ count = 0, params, setPagination
...
       page={currentPage}
-      onPerPageSelect={(_event: any, perPage: number, page: number) => {
+      onPerPageSelect={(_event: unknown, perPage: number, page: number) => {
         setPagination(returnOffsetVal(page), perPage);
       }}
-      onSetPage={(_event: any, page: number) => {
+      onSetPage={(_event: unknown, page: number) => {
         setPagination(returnOffsetVal(page));
       }}
       {...props}
```
Comment (jlmitch5) [inline src/Components/Pagination/Pagination.tsx:48]: more typescript refactorring I think

File: src/Containers/SavingsPlanner/Details/StatisticsTab/FormulaDescription/index.tsx [hunk 8]
```diff
@@ -1,37 +1,152 @@
...
+        <CostsText />
+        <br />
+
+        <SavingsText />
+      </Modal>
+    </>
+  );
+};
 
 export default FormulaDescription;
```
Comment (jlmitch5) [inline src/Containers/SavingsPlanner/Details/StatisticsTab/FormulaDescription/index.tsx:152]: Very nice, this component is very easy to understand!

File: src/Containers/SavingsPlanner/Details/StatisticsTab/TotalSavings/index.tsx [hunk 9]
```diff
@@ -20,9 +20,7 @@ const TotalSavings: FunctionComponent<Props> = ({
         size="4xl"
         style={{ color: isMoney ? '#81C46B' : '#0063CF' }}
       >
-        {isMoney
-          ? `${currencyFormatter(value)}`
-          : `${hoursFormatter(value)}`}
+        {isMoney ? `${currencyFormatter(value)}` : `${hoursFormatter(value)}`}
       </Title>
     </CardBody>
   </Card>
```
Comment (jlmitch5) [inline src/Containers/SavingsPlanner/Details/StatisticsTab/TotalSavings/index.tsx:26]: more refactoring

File: src/Containers/SavingsPlanner/Details/StatisticsTab/index.tsx (hunk 2 above)
Comment (jlmitch5) [reply src/Containers/SavingsPlanner/Details/StatisticsTab/index.tsx:327]: No problem

File: src/Containers/SavingsPlanner/Details/StatisticsTab/index.tsx [hunk 10]
```diff
@@ -251,59 +259,81 @@ const StatisticsTab: FunctionComponent<Props> = ({ tabsArray, data }) => {
...
+  tabsArray: PropTypes.arrayOf(
+    PropTypes.shape({
+      id: PropTypes.number.isRequired,
+      link: PropTypes.string.isRequired,
+      name: PropTypes.node.isRequired,
+    }).isRequired
+  ).isRequired,
+};
+
 export default StatisticsTab;
```
Comment (jlmitch5) [inline src/Containers/SavingsPlanner/Details/StatisticsTab/index.tsx:339]: This makes sense, I think it is 99% llinting-based refactoring

File: src/FeatureFlags/Provider.tsx [hunk 11]
```diff
@@ -1,25 +1,35 @@
...
 
-  return (<Provider value={features}>{children}</Provider>);
+  return <Provider value={features}>{children}</Provider>;
+};
+
+FeatureFlagProvider.propTypes = {
+  children: PropTypes.node.isRequired,
 };
 
 export default FeatureFlagProvider;
```
Comment (jlmitch5) [inline src/FeatureFlags/Provider.tsx:35]: again, I think ts linting refactoring

**Review by jlmitch5 — APPROVED** [review]
This is very good.  Like I mentioned, we probably could have made the savings plan formula stuff a couple files changes, and then had a separate pr which did the linting infrastucture and update of all the ts files so that it's clearer what new code was added, but it's all good.  Great job!

File: src/Api/api.ts (hunk 4 above)
Comment (brumik) [reply src/Api/api.ts:106]: Yes, I had to, it was all red and complaining and didn't compile

**Comment by brumik** [comment]
@jlmitch5 Yes, as I wrote in the PR description I pushed them unfortunately together the added linting and fix and the modal popup. Thank you for your review trough these bit messy PR.
